### ISSUE TITLE
Adding dataset from an organization page

### ISSUE THREAD

**Issue by anuveyatsu** [GSA/datagov-ckan-multi#476]
At the moment, there are 2 ways of navigating to new metadata form in the UI:

1. Go to "Dataset" page and click on "Add dataset" button (eg, https://inventory.sandbox.datagov.us/dataset)
2. Go to individual organization page and click on "Add dataset" button (eg, https://inventory.sandbox.datagov.us/organization/test-123)

We already can access the new metadata form using option 1, however, we cannot do it via option 2 because the template is overwritten in the `ckanext-datajson` extension here:

https://github.com/GSA/ckanext-datajson/blob/master/ckanext/datajson/templates/organization/read.html#L5

How should we fix it? We could try using jinja's "super" tag so that in the `datajson` extension we don't override entire `page_primary_action` block but only add some additional buttons. Read more about it here https://docs.ckan.org/en/2.5/theming/templates.html#extending-parent-blocks-with-jinja-s-super

**Comment by adborden** [comment]
I think we want a short-term fix to just update the ckanext-datajson extension to point to the new controller, a one-line change, right? I think that we will eventually replace/override the new dataset controller with our new one (new-metadata), in which case we could undo this one-line change to ckanext-datajson. Does that sound right?

If not, then I'm interested in doing whatever the CKAN best practice is. This certainly isn't the first time two extensions override the same template. `super` is okay, but I've seen all kinds of messes when two extensions need to override templates. I would prefer to avoid that.

**Comment by adborden** [comment]
... hmm, but we have to keep in mind that ckanext-datajson is also used by inventory and catalog today. We don't want to apply the change to catalog, and we only want this change to apply to inventory when we're ready to ship. Maybe we can put it behind a feature flag or configuration 🤷 .

**Comment by anuveyatsu** [comment]
@adborden OK, as I understand we have some requirements here:

* make sure that `datajson` extension works with both inventory and catalog apps as today;
* make sure we have clear docs about that implementation so that other developers can easily understand what we've done in the future.

Options that I think we could implement:

1. Refactor `datajson` extension so that it does only add new buttons so that we can easily modify "Add dataset" button in `dcat_usmetadata` extension.
2. Don't do any changes in `datajson` extension and try to modify "Add dataset" button in `dcat_usmetadata` extension. I can't guarantee that it would work but we can try to use `super` tag here and see. We will need to make sure that `dcat_usmetadata` extension always comes **after** `datajson` extension in the configuration file (in the list of `plugins`).

Any thoughts on this?

**Comment by anuveyatsu** [comment]
## Option 2

We've started with options 2 and we didn't succeed because `datajson` extension replaces the entire `page_primary_action` block of the organization page so it never reaches the template in the `dcat_usmetadata` extension.

Next, we've started thinking about other possible solutions which you can see below.

## Option 3

We have considered how `organization/read.html` template is implemented in the latest version of CKAN (eg, v2.9):

```
{% extends "organization/read_base.html" %}

{% block page_primary_action %}
  {% if h.check_access('package_create', {'owner_org': group_dict.id}) %}
    {% snippet 'snippets/add_dataset.html', group=group_dict.id %}
  {% endif %}
{% endblock %}

...
```

As you can see it allows developers to override the snippet (using the `snippet` tag ) for just "Add dataset" button. We thought we could implement this approach in the `datajson` extension so that we can replace it from the `dcat_usmetadata` extension. However, this approach didn't work again due to the similar reason as in option 2. It looks like we can only override the core CKAN templates.

This is what we did in the `datajson` extension (`organization/read.html`):

```
{% ckan_extends %}

{% block page_primary_action %}
    {% if h.check_access('package_create', {'owner_org': c.group_dict.id}) %}
    
        {% snippet 'snippets/add_dataset.html', group=group_dict.id %}

        {% if h.datajson_inventory_links_enabled() %}
        <button id="btnUnredacted" class="btn btn-secondary">Unredacted Inventory</button>
        <button id="btnRedacted" class="btn btn-secondary">Redacted Inventory</button>
        <button id="btnDraft" class="btn btn-secondary">Export Drafts</button>

        <form id="formUnredacted" action="{{c.group_dict.id + '/unredacted.json' }}"></form>
        <form id="formRedacted" action="{{c.group_dict.id + '/redacted.json' }}"></form>
        <form id="formDraft" action="{{c.group_dict.id + '/draft.json' }}"></form>

        <script type="text/javascript" src="/base/vendor/jquery.min.js"></script>
        <script type="text/javascript">
            $(document).ready(function () {
                $('#btnUnredacted').click(function () {
                    $('#formUnredacted').submit();
                });

                $('#btnRedacted').click(function () {
                    $('#formRedacted').submit();
                });

                $('#btnDraft').click(function () {
                    $('#formDraft').submit();
                });
            });
        </script>
        {% endif %}
    {% endif %}
{% endblock %}
```

## Option 4

Can we introduce brand new block in "DataJSON" extension, eg, `datajson_action` block, so that we don't change the primary action block?

This approach could work, however, there is no pattern showing how to add the new block into the specific part of the page. For example, we would need the new block (e.g. `datajson_action`)  to be placed right after the `page_primary_action` block. **So we haven't tried this approach.**

## Option 1

Finally, we decided to go with option 1 so that in `datajson` extension, we only **extend** the core template but not replace. You can see what has changed in this commit (it's in my fork of `datajson`):

https://github.com/anuveyatsu/ckanext-datajson/commit/1e1b9b32b53a531308c8e0de71c81c3c151cb60d

As you can see this wouldn't break any other apps using `datajson` extension. 
You might wonder what the original `page_primary_action` block looks like:

https://github.com/ckan/ckan/blob/release-v2.5.2/ckan/templates/organization/read.html#L3-L7

Next, we applied changes in `dcat_usmetadata` extnsion to add a new button to new metadata form (also `organization/read.html` file):

```
{% ckan_extends %}

{% block page_primary_action %}
    {% if h.check_access('package_create', {'owner_org': c.group_dict.id}) %}
        <a class="btn btn-primary" href="/dataset/new-metadata?group={{ c.group_dict.id }}">{{ _('Add Dataset') }}</a>
    {% endif %}

    {{ super() }}

{% endblock %}
```

![](https://i.imgur.com/oeueAfL.png)


As a result the button was added, however, we had original "Add dataset" button as well so we decided to hide it using JS for now:

```
<script type="text/javascript">
    jQuery(document).ready(function () {
        jQuery('.btn-primary').eq(1).hide();
    });
</script>
```

Final result:

![](https://i.imgur.com/CfuWjqc.png)

**QUESTION: if option 1 is ok by you, i will raise a PR to update the `datajson` extension and ask you for a formal review, please confirm?**

**Comment by adborden** [comment]
Is it possible to just override the `new` controller with the new metadata form? Since we're pinning the extension version on master, we no longer have to support both forms/controllers.

Otherwise Option 1 is fine. Can we move the button hiding to CSS?

**Comment by anuveyatsu** [comment]
@adborden my understanding was that we want to keep access to both forms (old and new) so we needed to have a temporary `new-metadata` controller. I believe it's best for us to keep access to both forms for now, not only for external users, but also for the scrum team to easily access the old form and compare how things were Vs how things are now in terms of features. Once, we are satisfied that we have ported most/all features from the old form to the new, then I believe we should override `new` controller and we'll remove the temporary solution.

Yes, we also could use CSS instead of JS for hiding the extra button.
